### PR TITLE
Change trait RangeArgument to struct RangeBounds.

### PR DESCRIPTION
**Pull request by clarfonthey**
This follows from the discussion in #30877. This removes the `RangeArgument` trait in favour of the `RangeBounds` struct, where `RangeArgument<T>` is loosely translated to `Into<RangeBounds<T>>`.

I'm still working on testing this but I'll upload the PR now and see how people feel.

Additionally, I was going to at some point refactor the code for dealing with ranges of `usize` into methods on `RangeBounds<usize>`. That isn't blocking on this commit, but it's something that should be done at some point.

### CODE REVIEW

**Comment by aidanhs** [comment]
Hi @clarcharr, looks like there has been a bunch of CI failures like
```
[00:01:59] error[E0549]: rustc_deprecated attribute must be paired with either stable or unstable attribute
[00:01:59]    --> /checkout/src/liballoc/lib.rs:214:1
[00:01:59]     |
[00:01:59] 214 | pub use core::ops::Bound;
[00:01:59]     | ^^^^^^^^^^^^^^^^^^^^^^^^^
[00:01:59] 
[00:01:59] error: use of unstable library feature 'range_argument' (see issue #30877)
[00:01:59]   --> /checkout/src/liballoc/btree/map.rs:16:24
[00:01:59]    |
[00:01:59] 16 | use core::ops::{Index, RangeBounds};
[00:01:59]    |                        ^^^^^^^^^^^
[00:01:59]    |
[00:01:59]    = help: add #![feature(range_argument)] to the crate attributes to enable
```

Is this PR still a work in progress (per your original comment)?

**Comment by clarcharr** [comment]
It is! I'll get back to this today or tomorrow.

**Comment by brson** [comment]
This needs to land within a week or not at all. It is slated for stabilization on 7/20.

I am not crazy about making API changes at the last minute as this does. A rename - ok, but changing how the API works seems unnecessarily risky.

The stabilization of this API is in [this backport](https://github.com/rust-lang/rust/pull/43114).

**Comment by durka** [comment]
I'm a bit confused about the API change, honestly. I see it was suggested https://github.com/rust-lang/rust/issues/30877#issuecomment-305834824 but there wasn't much discussion. There was consensus around a rename though, but it seems weird to rush the other change. Maybe there is more context that I missed?

**Comment by durka** [comment]
OK, I looked a little more closely at the situation.

1. We decided in the tracking issue to stabilize `RangeArgument`, and pretty much everyone agreed to change the name but nobody got around to doing it
1. During FCP, @johncf suggested the "`Into` trick".
1. Right after stabilization, @glaebhoerl complained about the naming.

This PR changes the name and the `Into` trick but does not stabilize anything. The backport stabilizes the API without either of those.

Based on that contradiction it seems to me that we haven't agreed on the design of this feature. Should we just revert the stabilization for now?

**Comment by sfackler** [comment]
Nothing related to this is being stabilized AFAIK.

**Comment by clarcharr** [comment]
@brson I should clarify that this is absolutely not expected to stabilise soon. The intention was to offer an implementation of the `Into` version as a POC to make sure that people are okay with it, let the dust settle, then stabilise it. Right now I still have a bit of work to do making `BTree*::range` work as expected, but I do feel like this is a better way forward.

If the team would prefer to leave `RangeArgument` as it is and simply rename it for the sake of getting this stabilised quicker, I can modify this PR to do that.

**Comment by liigo** [comment]
@durka  
> Based on that contradiction it seems to me that we haven't agreed on the design of this feature. Should we just revert the stabilization for now?

I think so.

**Comment by Mark-Simulacrum** [comment]
So I'm not clear what the status of this is. I don't think we're backporting this to beta at this point (it's too late).

**Comment by sfackler** [comment]
There's no need to beta backport this - RangeArgument was un-stabilized.

This does seem like a reasonable way forward though. r=me with a rebase!

File: src/liballoc/lib.rs [hunk 1]
```diff
@@ -202,64 +202,17 @@ mod std {
...
-}
-
 /// An intermediate trait for specialization of `Extend`.
 #[doc(hidden)]
 trait SpecExtend<I: IntoIterator> {
     /// Extends `self` with the contents of the given iterator.
     fn spec_extend(&mut self, iter: I);
 }
 
+#[rustc_deprecated(reason = "moved to core::ops", since = "1.19.0")]
```
Comment (ollie27) [inline src/liballoc/lib.rs]: This would need to be "1.21.0" now.

File: src/liballoc/range.rs [hunk 2]
```diff
@@ -11,142 +11,7 @@
 #![unstable(feature = "collections_range",
             reason = "waiting for dust to settle on inclusive ranges",
             issue = "30877")]
+#![rustc_deprecated(reason = "moved to core::ops", since = "1.19.0")]
```
Comment (ollie27) [inline src/liballoc/range.rs]: As this module only contained `RangeArgument` and that's been removed, this entire module can just be deleted along with its reexports in `collections` and `std::collections`.

Comment (clarcharr) [inline src/liballoc/range.rs]: Will do!

File: src/liballoc/lib.rs (hunk 1 above)
Comment (clarcharr) [inline src/liballoc/lib.rs]: Will do!

**Comment by alexcrichton** [comment]
pign @clarcharr, just wanted to make sure this stayed on your radar

**Comment by aidanhs** [comment]
@clarcharr I'm going to close this to keep the queue clean, but if you manage to get back to this to rebase let us know and we'll reopen!

**Pull request closed without merging**